### ISSUE TITLE
[Issue]: batched_gemm_a16wfp4 split-K path reads out of bounds / returns garbage for small M on gfx950

### ISSUE THREAD

**Issue by Concurrensee** [ROCm/aiter#3737]
### Problem Description

(Content summarized by AI, but verified)

## Summary

We found a correctness and stability issue in AITER's `batched_gemm_a16wfp4` kernel on ROCm/gfx950. The issue appears in the split-K path selected for small `M`, which is common during LLM decode. The kernel can return garbage values or hit a HIP illegal memory access.

This surfaced while running GLM-5.1 in vLLM, but a standalone test of the AITER kernel reproduces the issue without the full model. Disabling vLLM's use of this fp4 batched-GEMM path avoids the problem.

## Environment

- GPU arch: gfx950
- Container: `rocm/vllm-dev:nightly`
- AITER version: `0.1.13.post1`
- vLLM image version used during initial repro: `0.22.1rc1.dev511+gc621af169.rocm723`

## Problem Shape

The GLM-5.1 MLA decode up-projection uses shapes equivalent to:

- `x`: `[8, M, 512]`, bf16
- `w`: `[8, 256, 256]`, uint8
- `w_scales`: `[8, 256, 16]`, uint8
- `transpose_bm=True`
- `prequant=True`

For small `M`, `_get_config(M, N=256, K=512)` selects `NUM_KSPLIT=4`.

## Observed Behavior

Standalone kernel test results against a bf16 reference:

- `M = 1, 2, 4, 8, 16, 32, 64, 128`: output is garbage, cosine similarity near 0 or NaN
- `M = 256, 512`: output is correct, cosine similarity around 0.987
- In the full GLM-5.1 decode path, this sometimes becomes `HIP error 700: illegal memory access`

Forcing `NUM_KSPLIT=1` makes the small-M cases correct.

## Suspected Root Cause

For `M=1, N=256, K=512`, AITER selects:

- `NUM_KSPLIT=4`
- `BLOCK_SIZE_K=256`
- `SPLITK_BLOCK_SIZE=256`

The effective A ranges are:

- `pid_k=0`: `[0, 256)` valid
- `pid_k=1`: `[256, 512)` valid
- `pid_k=2`: `[512, 768)` out of bounds
- `pid_k=3`: `[768, 1024)` out of bounds

Only 2 splits are needed for `K=512`, but 4 are launched. The guard appears to use packed half-units:

### Operating System

22.04.5 LTS (Jammy Jellyfish)

### CPU

AMD EPYC 9575F 64-Core Processor

### GPU

8 * MI355

### ROCm Version

rocm 7.2.3

### Steps to Reproduce

```
"""Standalone test of AITER batched_gemm_a16wfp4 with GLM-5.1 MLA _v_up_proj
shapes, mirroring vllm/model_executor/layers/attention/mla_attention.py.

No model load: directly exercises the AITER fp4 batched-GEMM kernel that
faulted (illegal memory access in _batched_gemm_a16wfp4_reduce_kernel) to
determine whether the kernel itself is buggy independent of vLLM.

Run on a single GPU:
  AMD_SERIALIZE_KERNEL=3 HIP_LAUNCH_BLOCKING=1 VLLM_ROCM_USE_AITER=1 \
  HIP_VISIBLE_DEVICES=0 python3 /root/workspace/aiter_fp4bmm_test.py
"""

import os
import sys
import traceback

import torch

from vllm._aiter_ops import rocm_aiter_ops
from vllm.model_executor.layers.quantization.quark.utils import (
    quark_quantize_weight_to_mxfp4,
)

# GLM-5.1 MLA dims, per TP=8 rank.
N = int(os.environ.get("N_HEADS", "8"))        # num_heads / TP
LKV = int(os.environ.get("LKV", "512"))        # kv_lora_rank (K)
V = int(os.environ.get("V_HEAD", "256"))       # v_head_dim (N)

dev = "cuda"
print(f">>> fp4bmm_enabled={rocm_aiter_ops.is_fp4bmm_enabled()} "
      f"N={N} LKV={LKV} V={V}", flush=True)

# W_UV is [Lkv, N, V]; vLLM stores W_V = quant(W_UV.permute(1,2,0)) -> [N, V, Lkv]
W_UV = torch.randn(LKV, N, V, dtype=torch.bfloat16, device=dev) * 0.02
W_V, W_V_scale = quark_quantize_weight_to_mxfp4(W_UV.permute(1, 2, 0).contiguous())
print(f">>> W_V {tuple(W_V.shape)} {W_V.dtype}  scale {tuple(W_V_scale.shape)} "
      f"{W_V_scale.dtype}", flush=True)

batch_sizes = [int(b) for b in os.environ.get(
    "BATCHES", "1,2,4,8,16,32,64,128,256,512").split(",")]

def cossim(a, b):
    a = a.float().flatten()
    b = b.float().flatten()
    return torch.nn.functional.cosine_similarity(a, b, dim=0).item()


for B in batch_sizes:
    try:
        x = torch.randn(N, B, LKV, dtype=torch.bfloat16, device=dev) * 0.1
        out = torch.empty(B, N, V, dtype=torch.bfloat16, device=dev)
        out = out.view(-1, N, V)
        res = rocm_aiter_ops.batched_gemm_a16wfp4(
            x, W_V, W_V_scale, out,
            transpose_bm=True, prequant=True, y_scale=None,
        )
        torch.cuda.synchronize()
        # bf16 reference: out_ref[b,n,v] = sum_l x[n,b,l] * W_UV[l,n,v]
        ref = torch.einsum("nbl,lnv->bnv", x.float(), W_UV.float())
        finite = torch.isfinite(res).all().item()
        cs = cossim(res, ref)
        flag = "OK " if (finite and cs > 0.9) else "BAD"
        print(f"B={B:5d}  {flag} out={tuple(res.shape)} finite={finite} "
              f"cos_sim_vs_bf16={cs:.4f} mean={res.float().mean().item():.3e}",
              flush=True)
    except Exception as e:  # noqa: BLE001
        print(f"B={B:5d}  FAIL  {type(e).__name__}: {e}", flush=True)
        traceback.print_exc()
        sys.exit(2)

print(">>> ALL BATCHES DONE", flush=True)
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

**Comment by srinivamd** [comment]
## Root Cause Analysis

The `batched_gemm_a16wfp4` split-K path has three compounding defects that produce garbage output or `HIP error 700` for small M on gfx950:

1. **Missing `NUM_KSPLIT` re-ensure in kernel-local `get_splitk()`** — The copy at `_triton_kernels/gemm/batched/batched_gemm_a16wfp4.py` was missing the final `NUM_KSPLIT = triton.cdiv(K, (SPLITK_BLOCK_SIZE // 2))` recalculation that the canonical version at `gemm_afp4wfp4.py:63` has. Without this, `get_splitk` can return a `NUM_KSPLIT` that exceeds `ceil(K / (SPLITK_BLOCK_SIZE // 2))`, launching excess K-splits whose starting offset falls outside the valid K range.

2. **Missing `other=0.0` on masked load in reduction kernel** — When `ACTUAL_KSPLIT < MAX_KSPLIT`, the reduction kernel uses a masked `tl.load` for the `y_pp` partial-results buffer, but without `other=0.0`. Unexecuted K-splits leave their `y_pp` slots uninitialized (`torch.empty`), and the masked-off loads can sum garbage into the final output.

3. **Missing `NUM_KSPLIT=1` safety clamp in batched wrapper** — When `BLOCK_SIZE_K >= 2*K` (a single block covers all of K), the non-batched `gemm_a16wfp4.py` forces `config["NUM_KSPLIT"] = 1`. The batched wrapper was missing this clamp.

The reporter's AITER `0.1.13.post1` predates the config workaround in [ROCm/aiter PR #3283](https://github.com/ROCm/aiter/pull/3283) (merged 2026-05-28) which sets `NUM_KSPLIT=1` for all gfx950 configs — avoiding the buggy codepath via config but not fixing the kernel.

---

## GitHub Findings

### ROCm/aiter

| PR/Issue | Title | Date | Relevance | Confidence |
|---|---|---|---|---|
| [#3737](https://github.com/ROCm/aiter/issues/3737) | batched_gemm_a16wfp4 split-K OOB / garbage for small M on gfx950 | 2026-06-15 | Direct bug report | High |
| [#3766](https://github.com/ROCm/aiter/pull/3766) | Fix batched_gemm_a16wfp4 split-K garbage output / OOB for small M | 2026-06-17 | Root cause fix (3 defects) | High |
| [#3283](https://github.com/ROCm/aiter/pull/3283) | Tune gfx950 batched GEMM configs and benchmarks | 2026-05-28 | Config workaround (NUM_KSPLIT=1) | High |

### vllm-project/vllm

| PR/Issue | Title | Date | Relevance | Confidence |
|---|---|---|---|---|
| [vllm-project/vllm#20419](https://github.com/vllm-project/vllm/pull/20419) | Add CompressedTensorsMLAROCmImpl with FP4 batched GEMM | — | Code path that calls the buggy kernel | High |
| [vllm-project/vllm#34641](https://github.com/vllm-project/vllm/issues/34641) | Default FP4BMM=True crashes on MI300X (gfx942) | 2026-05-28 | Related — led to gfx950 hardware gate | Medium |

### ROCm/TheRock, ROCm/rocm-libraries, ROCm/rocm-systems

No matches found.

---

## Recommended Action

- [ ] **Merge fix PR** [#3766](https://github.com/ROCm/aiter/pull/3766) — addresses all three kernel defects
- [ ] **Immediate workaround** — set `VLLM_ROCM_USE_AITER_FP4BMM=0` or upgrade AITER past [#3283](https://github.com/ROCm/aiter/pull/3283)
- [ ] **Add regression tests** for small-M × small-K shapes with `NUM_KSPLIT > 1`

*Analysis generated with AI assistance.*
